Micronaut 4.1.x: the IdleStateEvent of the last request may cause a ReadTimeoutException on a healthy new request

### Expected Behavior

Consider a reactive client with a connection pool and some read timeout configured:
```
micronaut.http.client.read-timeout=2s
micronaut.http.client.pool.enabled=true
```

**Expected behavior:** Two consecutive requests (_A_ and _B_) that are executed approximately `2s` apart, should result in two successful responses.

![micronaut4issue](https://github.com/micronaut-projects/micronaut-core/assets/3586560/838d8b39-d066-4912-86f2-06d107a0e09b)


### Actual Behaviour

Instead, the second one produces a ReadTimeoutException:

![micronaut4issue2](https://github.com/micronaut-projects/micronaut-core/assets/3586560/673aa651-880c-49ed-a879-6e4cd7d8cfcf)


Note that requests _A_ and _B_ run faster than `2s` (around 110ms each), so this is not a performance error. This only happens when the delay between _A_ and _B_ is near `micronaut.http.client.read-timeout`. If you make the delay shorter (200ms, for example) or longer (3s), all the requests are OK again.

Apparently, there exists a time window (that starts approximately `2s` after _A_) in which a scheduled `IdleStateEvent` of _A_ will be understood as a timeout on _B_. This time window seems to be large enough to reproduce it easily with a simple sleep.

![micronaut4issue3](https://github.com/micronaut-projects/micronaut-core/assets/3586560/bd5ecf6f-2cf6-4856-a310-af75767d810c)


Stack trace of the ReadTimeoutException:
```
i.m.http.server.RouteExecutor - Unexpected error occurred: Read Timeout
io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
	at io.micronaut.http.client.exceptions.ReadTimeoutException.<clinit>(ReadTimeoutException.java:26)
	at io.micronaut.http.client.netty.DefaultHttpClient$BaseHttpResponseHandler.exceptionCaught(DefaultHttpClient.java:1996)
	at io.micronaut.http.client.netty.DefaultHttpClient$FullHttpResponseHandler.exceptionCaught(DefaultHttpClient.java:2267)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.exceptionCaught(DefaultChannelPipeline.java:1377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.DefaultChannelPipeline.fireExceptionCaught(DefaultChannelPipeline.java:907)
	at io.netty.handler.codec.http2.Http2MultiplexHandler$2.visit(Http2MultiplexHandler.java:314)
	at io.netty.handler.codec.http2.Http2FrameCodec$1.visit(Http2FrameCodec.java:197)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$ActiveStreams.forEachActiveStream(DefaultHttp2Connection.java:978)
	at io.netty.handler.codec.http2.DefaultHttp2Connection.forEachActiveStream(DefaultHttp2Connection.java:209)
	at io.netty.handler.codec.http2.Http2FrameCodec.forEachActiveStream(Http2FrameCodec.java:193)
	at io.netty.handler.codec.http2.Http2ChannelDuplexHandler.forEachActiveStream(Http2ChannelDuplexHandler.java:83)
	at io.netty.handler.codec.http2.Http2MultiplexHandler.fireExceptionCaughtForActiveStream(Http2MultiplexHandler.java:309)
	at io.netty.handler.codec.http2.Http2MultiplexHandler.exceptionCaught(Http2MultiplexHandler.java:298)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:143)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.exceptionCaught(Http2ConnectionHandler.java:582)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.handler.ssl.SslHandler.exceptionCaught(SslHandler.java:1115)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.exceptionCaught(DefaultChannelPipeline.java:1377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.DefaultChannelPipeline.fireExceptionCaught(DefaultChannelPipeline.java:907)
	at io.micronaut.http.client.netty.ConnectionManager$Pool$Http2ConnectionHolder.fireReadTimeout(ConnectionManager.java:1350)
	at io.micronaut.http.client.netty.ConnectionManager$Pool$ConnectionHolder$1.readTimedOut(ConnectionManager.java:1088)
	at io.netty.handler.timeout.ReadTimeoutHandler.channelIdle(ReadTimeoutHandler.java:90)
	at io.netty.handler.timeout.IdleStateHandler$ReaderIdleTimeoutTask.run(IdleStateHandler.java:503)
	at io.netty.handler.timeout.IdleStateHandler$AbstractIdleTask.run(IdleStateHandler.java:475)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

#### Possible cause of the error:
We believe that the cause of the error is that a `IdleStateEvent` of _A_ is handled here:
https://github.com/micronaut-projects/micronaut-core/blame/4.1.x/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java#L1085

And the `if (hasLiveRequests())` resolves to true, but not due _A_ being timeout, but due the new request _B_ is being handled in the ConnectionPool in the same precise moment.

#### Other notes:
- This issue doesn't reproduce on Micronaut 3.x.x. This is keeping us from using 4.x.x right now.
- Apparently this happens more often than we could expect. In a production environment we had a 3% error rate due to this timeout.


### Steps To Reproduce

1. Use Java 17.
2. Clone [the given repository](https://github.com/alsoba13/micronaut4-read-timeout-in-connection-pool-issue).
3. Follow one of the two given options to reproduce. Example of reproduction using `./gradlew test --info`:
      ```
      ReproduceIssue > runConsecutiveRequests() STANDARD_OUT

      - Running 10 request with 100ms delay between requests

        (874ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (157ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (164ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (173ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (162ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (161ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (161ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (170ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (171ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --
        (202ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 100ms --

      - Running 10 request with 1900ms delay between requests

        (163ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 1900ms --
        (156ms) 500 - io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
        -- Sleeping for 1900ms --
        (609ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 1900ms --
        (99ms) 500 - io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
        -- Sleeping for 1900ms --
        (683ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 1900ms --
        (95ms) 500 - io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
        -- Sleeping for 1900ms --
        (663ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 1900ms --
        (101ms) 500 - io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
        -- Sleeping for 1900ms --
        (649ms) 200 - Micronaut launch page has 3118 characters
        -- Sleeping for 1900ms --
        (101ms) 500 - io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
        -- Sleeping for 1900ms --
      ```

### Environment Information

- macOs 13.2 (22D49)
- java 17.0.7

### Example Application

https://github.com/alsoba13/micronaut4-read-timeout-in-connection-pool-issue

### Version

4.1.x

We are facing the same issue on a low-request service

```
Caused by: io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout
	at io.micronaut.http.client.exceptions.ReadTimeoutException.<clinit>(ReadTimeoutException.java:26)
	at io.micronaut.http.client.netty.DefaultHttpClient$BaseHttpResponseHandler.exceptionCaught(DefaultHttpClient.java:2045)
	at io.micronaut.http.client.netty.DefaultHttpClient$FullHttpResponseHandler.exceptionCaught(DefaultHttpClient.java:2279)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireExceptionCaught(CombinedChannelDuplexHandler.java:424)
	at io.netty.channel.ChannelHandlerAdapter.exceptionCaught(ChannelHandlerAdapter.java:92)
	at io.netty.channel.CombinedChannelDuplexHandler$1.fireExceptionCaught(CombinedChannelDuplexHandler.java:145)
	at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:143)
	at io.netty.channel.CombinedChannelDuplexHandler.exceptionCaught(CombinedChannelDuplexHandler.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.handler.ssl.SslHandler.exceptionCaught(SslHandler.java:1198)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.micronaut.http.client.netty.ConnectionManager$Pool$Http1ConnectionHolder.fireReadTimeout(ConnectionManager.java:1244)
	at io.micronaut.http.client.netty.ConnectionManager$Pool$ConnectionHolder$1.readTimedOut(ConnectionManager.java:1086)
	at io.netty.handler.timeout.ReadTimeoutHandler.channelIdle(ReadTimeoutHandler.java:90)
	at io.netty.handler.timeout.IdleStateHandler$ReaderIdleTimeoutTask.run(IdleStateHandler.java:503)
	at io.netty.handler.timeout.IdleStateHandler$AbstractIdleTask.run(IdleStateHandler.java:475)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 common frames omitted
```

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ConnectionManager.java
@@ -1064,9 +1064,20 @@ public class ConnectionManager {
             ScheduledFuture<?> ttlFuture;
             volatile boolean windDownConnection = false;
 
+            private ResettableReadTimeoutHandler readTimeoutHandler;
+
             ConnectionHolder(Channel channel, NettyClientCustomizer connectionCustomizer) {
                 this.channel = channel;
                 this.connectionCustomizer = connectionCustomizer;
+            }
+
+            /**
+             * Reset the read timeout, i.e. start it from 0 again.
+             */
+            private void resetReadTimeout() {
+                if (readTimeoutHandler != null) {
+                    readTimeoutHandler.resetReadTimeoutMn();
+                }
             }
 
             /**
@@ -1079,15 +1090,19 @@ public class ConnectionManager {
             final void addTimeoutHandlers(String before) {
                 // read timeout handles timeouts *during* a request
                 configuration.getReadTimeout()
-                    .ifPresent(dur -> channel.pipeline().addBefore(before, ChannelPipelineCustomizer.HANDLER_READ_TIMEOUT, new ReadTimeoutHandler(dur.toNanos(), TimeUnit.NANOSECONDS) {
-                        @Override
-                        protected void readTimedOut(ChannelHandlerContext ctx) {
-                            if (hasLiveRequests()) {
-                                fireReadTimeout(ctx);
-                                ctx.close();
+                    .ifPresent(dur -> {
+                        ResettableReadTimeoutHandler readTimeoutHandler = new ResettableReadTimeoutHandler(dur.toNanos(), TimeUnit.NANOSECONDS) {
+                            @Override
+                            protected void readTimedOut(ChannelHandlerContext ctx) {
+                                if (hasLiveRequests()) {
+                                    fireReadTimeout(ctx);
+                                    ctx.close();
+                                }
                             }
-                        }
-                    }));
+                        };
+                        this.readTimeoutHandler = readTimeoutHandler;
+                        channel.pipeline().addBefore(before, ChannelPipelineCustomizer.HANDLER_READ_TIMEOUT, readTimeoutHandler);
+                    });
                 // pool idle timeout happens *outside* a request
                 configuration.getConnectionPoolIdleTimeout()
                     .ifPresent(dur -> channel.pipeline().addBefore(before, ChannelPipelineCustomizer.HANDLER_IDLE_STATE, new ReadTimeoutHandler(dur.toNanos(), TimeUnit.NANOSECONDS) {
@@ -1160,9 +1175,13 @@ public class ConnectionManager {
                     return true;
                 }
                 if (channel.eventLoop().inEventLoop()) {
+                    resetReadTimeout();
                     dispatch0(sink);
                 } else {
-                    channel.eventLoop().execute(() -> dispatch0(sink));
+                    channel.eventLoop().execute(() -> {
+                        resetReadTimeout();
+                        dispatch0(sink);
+                    });
                 }
                 return true;
             }

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ResettableReadTimeoutHandler.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ResettableReadTimeoutHandler.java
@@ -20,6 +20,7 @@ import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
 import io.netty.handler.timeout.ReadTimeoutHandler;
 
+import java.util.NoSuchElementException;
 import java.util.concurrent.TimeUnit;
 
 /**
@@ -51,7 +52,10 @@ class ResettableReadTimeoutHandler extends ReadTimeoutHandler {
     @Override
     public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
         super.handlerRemoved(ctx);
-        ctx.pipeline().remove(ctx.name() + "-reset-interceptor");
+        try {
+            ctx.pipeline().remove(ctx.name() + "-reset-interceptor");
+        } catch (NoSuchElementException ignored) {
+        }
     }
 
     @Override

--- a/http-client/src/main/java/io/micronaut/http/client/netty/ResettableReadTimeoutHandler.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/ResettableReadTimeoutHandler.java
@@ -1,0 +1,91 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client.netty;
+
+import io.micronaut.core.annotation.Internal;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.timeout.ReadTimeoutHandler;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Resettable version of {@link ReadTimeoutHandler}, as a workaround before
+ * <a href="https://github.com/netty/netty/pull/13598">https://github.com/netty/netty/pull/13598</a>
+ * is merged. (TODO: move to new API when that is merged)
+ *
+ * @author Jonas Konrad
+ * @since 4.1.4
+ */
+@Internal
+class ResettableReadTimeoutHandler extends ReadTimeoutHandler {
+    private static final Object FAKE_MESSAGE = new Object();
+
+    private ChannelHandlerContext ctx;
+    private boolean reading = false;
+
+    public ResettableReadTimeoutHandler(long timeout, TimeUnit unit) {
+        super(timeout, unit);
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        super.handlerAdded(ctx);
+        this.ctx = ctx;
+        ctx.pipeline().addAfter(ctx.name(), ctx.name() + "-reset-interceptor", NextInterceptor.INSTANCE);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        super.handlerRemoved(ctx);
+        ctx.pipeline().remove(ctx.name() + "-reset-interceptor");
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        reading = true;
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        reading = false;
+        super.channelReadComplete(ctx);
+    }
+
+    void resetReadTimeoutMn() {
+        if (!reading) {
+            try {
+                channelRead(ctx, FAKE_MESSAGE);
+                channelReadComplete(ctx);
+            } catch (Exception ignored) {
+            }
+        }
+    }
+
+    @Sharable
+    private static class NextInterceptor extends ChannelInboundHandlerAdapter {
+        static final NextInterceptor INSTANCE = new NextInterceptor();
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            if (msg == FAKE_MESSAGE) {
+                return;
+            }
+            super.channelRead(ctx, msg);
+        }
+    }
+}

--- a/http-client/src/test/groovy/io/micronaut/http/client/netty/ConnectionManagerSpec.groovy
+++ b/http-client/src/test/groovy/io/micronaut/http/client/netty/ConnectionManagerSpec.groovy
@@ -564,6 +564,40 @@ class ConnectionManagerSpec extends Specification {
         ctx.close()
     }
 
+    def 'http1 read timeout during dispatch'() {
+        def ctx = ApplicationContext.run([
+                'micronaut.http.client.read-timeout': '5s',
+        ])
+        def client = ctx.getBean(DefaultHttpClient)
+
+        def conn = new EmbeddedTestConnectionHttp1()
+        conn.setupHttp1()
+        patch(client, conn)
+
+        // do one request
+        conn.testExchangeResponse(conn.testExchangeRequest(client))
+        conn.clientChannel.unfreezeTime()
+        // wait for one part of the interval
+        TimeUnit.SECONDS.sleep(2)
+        conn.advance()
+
+        // second request
+        def future = Mono.from(client.exchange('http://example.com/foo', String)).toFuture()
+        conn.advance()
+
+        // todo: move to advanceTime once IdleStateHandler supports it
+        // wait for the second part of the interval: below read-timeout, but together with the first sleep, above it
+        TimeUnit.SECONDS.sleep(3)
+        conn.advance()
+
+        assert !future.isDone()
+        conn.testExchangeResponse(future)
+
+        cleanup:
+        client.close()
+        ctx.close()
+    }
+
     def 'http1 ttl'() {
         def ctx = ApplicationContext.run([
                 'micronaut.http.client.connect-ttl': '100s',

